File: script/game_api_service.py
import requests

class GameApiService:
    @staticmethod
    def fetch_games(profile_id, since=None, until_game_id=None):
        api_url = f'https://aoe4world.com/api/v0/players/{profile_id}/games'
        params = {}
        if since:
            params['since'] = since
            
        all_games = []
        page = 1
        
        while True:
            params['page'] = page
            response = requests.get(api_url, params=params)
            if response.status_code != 200:
                print(f"Error fetching games: {response.status_code}")
                print(f"Response: {response.text}")
                break
                
            gamesResp = response.json()
            if not gamesResp:
                raise ValueError("No games found")

            games = gamesResp.get('games', [])
                
            print(f"Fetched {len(games)} games from page {page}.")
            print(f"Games: {games}")
            # If we're searching until a specific game, check if it's in this batch
            if until_game_id:
                game_ids = [game.get('id') for game in games]
                if until_game_id in game_ids:
                    until_index = game_ids.index(until_game_id)
                    # Only add games that come before the until_game_id
                    all_games.extend(games[:until_index])
                    break
            
            all_games.extend(games)
            page += 1
            total_count = gamesResp.get("total_count")
            offset = gamesResp.get("offset")
            count = gamesResp.get("count")
            if total_count == (offset + count):
                print("All games fetched.")
                break
        return all_games

File: back/game_api_service.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sync"
)

type GameApiService struct{}

// In-memory cache for games by profileID
var (
	gameCache      = make(map[int][]Game)
	gameCacheMutex sync.RWMutex
)

func (g *GameApiService) FetchGames(profileID int, since string, untilGameID string) ([]Game, error) {
	// Check cache first
	gameCacheMutex.RLock()
	cachedGames, found := gameCache[profileID]
	gameCacheMutex.RUnlock()
	if found {
		return cachedGames, nil
	}

	apiURL := fmt.Sprintf("https://aoe4world.com/api/v0/players/%d/games", profileID)
	allGames := []Game{}
	page := 1

	for {
		params := ""
		if since != "" {
			params += fmt.Sprintf("?since=%s", since)
		} else {
			params += "?"
		}
		params += fmt.Sprintf("&page=%d", page)

		resp, err := http.Get(apiURL + params)
		if err != nil {
			return nil, err
		}
		defer resp.Body.Close()
		if resp.StatusCode != 200 {
			return nil, fmt.Errorf("API error: %d", resp.StatusCode)
		}
		var gamesResp struct {
			Games      []Game `json:"games"`
			TotalCount int    `json:"total_count"`
			Offset     int    `json:"offset"`
			Count      int    `json:"count"`
		}
		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return nil, err
		}
		err = json.Unmarshal(body, &gamesResp)
		if err != nil {
			return nil, err
		}

		// If untilGameID is set, stop at that game
		if untilGameID != "" {
			found := false
			for i, game := range gamesResp.Games {
				if game.ID == untilGameID {
					allGames = append(allGames, gamesResp.Games[:i]...)
					found = true
					break
				}
			}
			if found {
				break
			}
		}

		allGames = append(allGames, gamesResp.Games...)

		// Check if we've fetched all games
		if gamesResp.TotalCount == (gamesResp.Offset+gamesResp.Count) || len(gamesResp.Games) == 0 {
			break
		}
		page++
	}
	gameCacheMutex.Lock()
	gameCache[profileID] = allGames
	gameCacheMutex.Unlock()

	return allGames, nil
}

func processPlayerGames(profileID int, filePath string) ([]Game, error) {
	ss := getStorageService()
	var existingGames []Game
	_ = ss.LoadData(filePath, &existingGames)

	var latestGameTime string
	if len(existingGames) > 0 {
		latestGameTime = existingGames[len(existingGames)-1].StartedAt
	}
	api := &GameApiService{}
	newGames, err := api.FetchGames(profileID, latestGameTime, "")
	if err != nil {
		return nil, err
	}
	allGames := append(existingGames, newGames...)
	err = ss.SaveData(filePath, &allGames)
	if err != nil {
		return nil, err
	}
	return allGames, nil
}
